### PR TITLE
Metric 4 Attribute Analysis, Prerun permutation test update

### PR DESCRIPTION
**Pull request by ryanzhang22**

### CODE REVIEW

File: Attribute Analysis.ipynb [hunk 1]
```diff
@@ -602,6 +603,340 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "import warnings\n",
```
Comment (ryanzhang22) [inline Attribute Analysis.ipynb:1008]: I added this here to suppress permutation test warnings about the number of items in a class being less than the number of splits.

File: prerun_analyzegen.py [hunk 2]
```diff
@@ -155,20 +154,15 @@ def country_to_iso3(country):
             b_indices = np.argsort(np.array(b_probs))
 
             pickle.dump([a_indices, b_indices, scene_filepaths[i], a_probs, b_probs], open("results/{0}/4/{1}_info.pkl".format(folder_name, names[categories[i]]), "wb"))
-            np.random.shuffle(projected_features_scenes)
-            clf_random.fit(projected_features_scenes, labels)
-            acc_random = clf_random.score(projected_features_scenes, labels)
-            value = acc / acc_random
+            
+            base_acc, rand_acc, p_value = permutation_test_score(clf, projected_features_scenes, labels, scoring="accuracy", n_permutations=100)
```
Comment (Angelina-Wang) [inline prerun_analyzegen.py:161]: I don't think you ever import permutation_test_score

File: Attribute Analysis.ipynb [hunk 3]
```diff
@@ -602,6 +603,340 @@
     "display(ui, out)"
    ]
   },
+  {
+   "source": [
+    "# (M4) Metric: Linearly separable objects by attribute\n",
```
Comment (Angelina-Wang) [inline Attribute Analysis.ipynb]: please adjust to leslie's new names

File: prerun_analyzegeo.py [hunk 4]
```diff
@@ -169,20 +168,20 @@ def sixprep(dataset, folder_name):
...
-            value = j_to_acc[diff_subregion] / acc_random # ratio between actual accuracy and accuracy of randomly shuffled labels
-            if value <= 1.2 and acc <= .7: # can tune as desired
+            def subregion_scoring(estimator, X_test, y_test):
+                y_pred = estimator.predict(X_test)
+                y_test[np.where(y_test!=diff_subregion)[0]] = -1
+                y_pred[np.where(y_pred!=diff_subregion)[0]] = -1
+                acc_random = np.mean(y_test == y_pred)
+                return acc_random
+
+            base_acc, rand_acc, p_value = permutation_test_score(clf, all_features, labels, scoring=subregion_scoring, n_permutations=100)
```
Comment (Angelina-Wang) [inline prerun_analyzegeo.py:178]: same, need to import this function

File: Attribute Analysis.ipynb [hunk 5]
```diff
@@ -970,6 +970,333 @@
...
+    "        amount = len(features_instances)\n",
+    "        phrase = [ratio, names[categories[i]], acc, p_value, len(features_instances), num_features]\n",
+    "        value_to_phrase[i] = phrase\n",
+    "\n",
+    "        for j in range(len(scene_classes)):\n",
+    "            dists = [scenes_per_attr[j][k] for k in range(num_attrs)]\n",
+    "            all_a = [np.zeros(len(scenes[i][k])) for k in range(num_attrs)]\n",
+    "            for d, a in zip(dists, all_a):\n",
+    "                a[:len(d)] = 1\n",
+    "            \n",
```
Comment (a-kleiman) [inline Attribute Analysis.ipynb]: Don't need a new space here

File: Attribute Analysis.ipynb [hunk 6]
```diff
@@ -970,6 +970,333 @@
...
+    "            for k in range(num_attrs):\n",
+    "                for l in range(k + 1, num_attrs):\n",
+    "                    _, p = stats.ttest_ind(all_a[k], all_a[l])\n",
+    "                    if not np.isnan(p):\n",
+    "                        value_to_scenephrase[p] = [names[categories[i]], scene_classes[j], len(dists[k]), len(all_a[k]), len(dists[l]), len(all_a[l]), attr_names[k], attr_names[l]]\n",
```
Comment (a-kleiman) [inline Attribute Analysis.ipynb:1163]: Add New line here for extra space

File: Attribute Analysis.ipynb [hunk 7]
```diff
@@ -970,6 +970,333 @@
...
+    "    pickle.dump([value_to_phrase, value_to_scenephrase], open(\"checkpoints/{}/att_clu.pkl\".format(folder_name), 'wb'))\n",
+    "else:\n",
```
Comment (anatk2020) [inline Attribute Analysis.ipynb:1151]: New line here

File: prerun_analyzegeo.py [hunk 8]
```diff
@@ -58,7 +59,7 @@ def country_to_iso3(country):
     return iso3
 
 def sixprep(dataset, folder_name):
-    info_stats = pickle.load(open("results/{}/geo_tag.pkl".format(folder_name), "rb")) #20GB
+    info_stats = pickle.load(open("results/{}/results/geo_tag.pkl".format(folder_name), "rb")) #20GB
```
Comment (Angelina-Wang) [inline prerun_analyzegeo.py]: is this correct? results/{}/results/geo_tag?

File: Attribute Analysis.ipynb [hunk 9]
```diff
@@ -26,9 +26,12 @@
     "2. att_siz Metric: [Distance from center, size, attribute label inference](#att_size)<br/>\n",
     "    2.1 [Setup](#att_size_setup)<br/>\n",
     "    2.2 [Analyses](#att_size_analyses)\n",
-    "3. att_cnt Metric: [Object occurrences and cooccurrences](#att_cnt))<br/>\n",
+    "3. att_cnt Metric: [Object occurrences and cooccurrences](#att_cnt)<br/>\n",
     "    3.1 [Setup](#att_cnt_setup)<br/>\n",
-    "    3.2 [Analyses](#att_cnt_analyses)"
+    "    3.2 [Analyses](#att_cnt_analyses)\n",
+    "4. att_clu Metric: [Linearly seperate objects by attribute](#att_clu)<br/>\n",
```
Comment (Angelina-Wang) [inline Attribute Analysis.ipynb]: can we make this number 5 to match gender analysis ordering

**Pull request merged**